### PR TITLE
Port CPU Trace from TH to ATen

### PR DESCRIPTION
**Pull request by zou3519**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#47126 Port CPU Trace from TH to ATen**
* #47125 Add test for trace

Context
-------
This PR is a rebase of @shihongzhi's https://github.com/pytorch/pytorch/pull/35360.
I forgot to merge it back when it was submitted so I rebased it and ran new benchmarks on it.

Benchmarks
----------

TL;DR: The op has more overhead than the TH version but for larger shapes the overhead disappears.

```
import torch

shapes = [
    [1, 1],
    [100, 100],
    [1000, 1000],
    [10000, 10000],
    [100000, 100000],
]

for shape in shapes:
    x = torch.ones(shape)
    %timeit x.trace()

Before:
1.83 µs ± 42.4 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)
1.98 µs ± 48.2 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
3.19 µs ± 10.7 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
85.2 µs ± 700 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
1.23 ms ± 4.34 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)

After:
2.16 µs ± 325 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)
2.08 µs ± 275 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
4.45 µs ± 19.2 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
81.8 µs ± 766 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
1.27 ms ± 6.75 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
```

Future work
-----------
Things that can be done after this PR:
- add complex tensor support
- Fix the type promotion discrepancy between CPU and CUDA

Differential Revision: [D24683259](https://our.internmc.facebook.com/intern/diff/D24683259)

### CODE REVIEW

File: aten/src/ATen/native/TensorShape.cpp [hunk 1]
```diff
@@ -1984,4 +1985,29 @@ Tensor movedim(const Tensor& self, int64_t src, int64_t dst) {
   return at::movedim(self, IntArrayRef{src}, IntArrayRef{dst});
 }
 
+Tensor trace_cpu(const Tensor& self) {
+  Tensor result = at::empty({}, self.options());
+  AT_DISPATCH_ALL_TYPES(self.scalar_type(), "trace", [&] {
```
Comment (anjali411) [inline aten/src/ATen/native/TensorShape.cpp:1990]: can we also enable it for complex dtypes? it should be straightforward to do that (similar to `torch.sum`)

Comment (zou3519) [reply aten/src/ATen/native/TensorShape.cpp:1990]: I took a quick look. Enabling it for complex dtypes isn't as simple as changing the dispatch macro here; we also have to change the CUDA trace implementation to work with complex.

The CUDA trace implementation requires that torch.diag() works with complex (it doesn't currently).

To keep things simple, I'd like to propose just shipping this as-is and leaving the complex enabling in a follow-up PR

**Review by anjali411 — APPROVED** [review]

File: aten/src/ATen/native/TensorShape.cpp [hunk 2]
```diff
@@ -1984,4 +1985,29 @@ Tensor movedim(const Tensor& self, int64_t src, int64_t dst) {
   return at::movedim(self, IntArrayRef{src}, IntArrayRef{dst});
 }
 
+Tensor trace_cpu(const Tensor& self) {
```
Comment (anjali411) [inline aten/src/ATen/native/TensorShape.cpp:1988]: @zou3519 I think we should update this to be a composite of `diag` and `sum` to remove the type promotion inconsistencies. Obviously, we should benchmark the performance and keep the existing implementation if it's favorable.

Comment (zou3519) [reply aten/src/ATen/native/TensorShape.cpp:1988]: I benchmarked the performance of diag() and sum() a while back. It turns out that diag() + sum() is slower than the original TH implementation which is why the code looks like how it does today :/

**Pull request closed without merging**